cli/ops/plugin.rs

```diff
@@ -151,3 +169,51 @@ impl Drop for PluginOpAsyncFuture {
     self.fut.take();
   }
 }
+
+/// A resource that has been inserted in the resource table by a dynamically
+/// loaded plugin.
+struct RefResource {
+  inner: Box<dyn Resource>,
+  _plugin_lib: Rc<Library>,
+}
+
+impl<'a> plugin_api::ResourceTable for PluginInterface<'a> {
+  fn add(&mut self, name: &str, resource: Box<dyn Resource>) -> ResourceId {
+    let ref_resource = RefResource {
+      inner: resource,
+      _plugin_lib: self.plugin_lib.clone(),
+    };
+    self.resource_table_mut().add(name, Box::new(ref_resource))
+  }
+
+  fn get(&self, rid: ResourceId) -> Option<Ref<dyn Resource>> {
+    // Resources inserted by plugins are wrapped in a RefResource wrapper, so
+    // we can ensure a reference to Rc<Library> is kept for as long as the
+    // resource stays in the resource table.
+    let table = self.resource_table();
+    match table.get::<RefResource>(rid) {
+      Some(_) => Some(Ref::map(table, |t| {
+        &*t.get::<RefResource>(rid).unwrap().inner
+      })),
+      None => None,
+    }
+  }
+
+  fn get_mut(&mut self, rid: ResourceId) -> Option<RefMut<dyn Resource>> {
+    let mut table = self.resource_table_mut();
+    match table.get_mut::<RefResource>(rid) {
+      Some(_) => Some(RefMut::map(table, |t| {
+        &mut *t.get_mut::<RefResource>(rid).unwrap().inner
+      })),
+      None => None,
+    }
+  }
+
+  fn remove(&mut self, rid: ResourceId) -> Option<Box<dyn Resource>> {
+    let mut table = self.resource_table_mut();
+    match table.get_mut::<RefResource>(rid) {
+      Some(_) => table.remove::<RefResource>(rid).map(|rc| rc.inner),
+      None => None, // `rid` does not reference a RefResource, so leave it.
```

Comment:
Well... that was not necessarily the intention here, but in the old code, the the resource was not a RefResource, it would *remove* the resource and then return *None*. That just seemed wrong.

Reply:
I made it so that 'close()' and 'has()' work for all types of resources.
For get()/get_mut() that won't be possible because the actual ResourceTable doesn't support that.

Reply:
This comment also applies to `get`/`get_mut`. Didn't really intend to be specific to `remove` it was just the only place where this fact is stated in a comment.